File: backend/app/schemas/handover_requests.py
import enum
from datetime import datetime
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, StringConstraints

from app.models.handover_requests import HandoverRequestStatus


class HandoverCandidateDecision(enum.StrEnum):
    confirm = "confirm"
    reject = "reject"


class HandoverReviewDecision(enum.StrEnum):
    approve = "approve"
    reject = "reject"


ReasonText = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=500)]


class HandoverRequestCreate(BaseModel):
    sub_project_ids: list[UUID] = Field(min_length=1)
    to_user_id: UUID
    reason: ReasonText

    model_config = ConfigDict(extra="forbid")


class HandoverCandidateReview(BaseModel):
    decision: HandoverCandidateDecision
    comment: str | None = None

    model_config = ConfigDict(extra="forbid")


class HandoverReviewRequest(BaseModel):
    decision: HandoverReviewDecision
    review_comment: str | None = None

    model_config = ConfigDict(extra="forbid")


class HandoverRequestRead(BaseModel):
    id: UUID
    from_user_id: UUID
    to_user_id: UUID
    sub_project_ids: list[UUID]
    reason: str
    status: HandoverRequestStatus
    candidate_comment: str | None
    candidate_responded_at: datetime | None
    reviewer_id: UUID | None
    review_comment: str | None
    reviewed_at: datetime | None
    forced_by_id: UUID | None
    forced_at: datetime | None
    created_at: datetime
    updated_at: datetime


class HandoverRequestListRead(BaseModel):
    items: list[HandoverRequestRead]
    total: int

File: frontend/src/types/handoverRequests.ts
export type HandoverRequestStatus =
  | 'pending_candidate'
  | 'candidate_rejected'
  | 'pending_review'
  | 'review_rejected'
  | 'approved'
  | 'forced';

export type HandoverCandidateDecision = 'confirm' | 'reject';
export type HandoverReviewDecision = 'approve' | 'reject';

export interface HandoverRequestRead {
  candidate_comment: string | null;
  candidate_responded_at: string | null;
  created_at: string;
  forced_at: string | null;
  forced_by_id: string | null;
  from_user_id: string;
  id: string;
  reason: string;
  review_comment: string | null;
  reviewed_at: string | null;
  reviewer_id: string | null;
  status: HandoverRequestStatus;
  sub_project_ids: readonly string[];
  to_user_id: string;
  updated_at: string;
}

export interface HandoverRequestListRead {
  items: HandoverRequestRead[];
  total: number;
}

export interface HandoverRequestListQuery {
  status?: HandoverRequestStatus;
}

export interface HandoverRequestCreatePayload {
  reason: string;
  subProjectIds: string[];
  toUserId: string;
}

export interface HandoverCandidateReviewPayload {
  comment?: string | null;
  decision: HandoverCandidateDecision;
}

export interface HandoverReviewPayload {
  decision: HandoverReviewDecision;
  reviewComment?: string | null;
}

export const HANDOVER_REQUEST_STATUS_LABELS: Record<HandoverRequestStatus, string> = {
  approved: '已通过',
  candidate_rejected: '候选人已拒绝',
  forced: '已强制转交',
  pending_candidate: '待候选人确认',
  pending_review: '待审核',
  review_rejected: '审核已驳回',
};
